### PR TITLE
Store metadata in WAL

### PR DESCRIPTION
**Pull request by tpaschalis**
Signed-off-by: Paschalis Tsilias <paschalist0@gmail.com>

## Changes
This PR updates Prometheus to store metadata in its Write-Ahead Log. To achieve this, a new `MetadataAppender` interface is introduced and embedded in Appender.

Its method implementation is used to store a `RefMetadata` slice as a headAppender field, which is then encoded in the WAL according to the proposed format.

The scrape loop is also modified to detect changes in metadata in both new and old (cached) series and pass them along.

To verify that metadata is indeed appended to the WAL, I use a commit such as tpaschalis/prometheus@fd36aabb3fff2ead259c6f559f3c9b3020f107a9 to check and print metadata in the loadWAL method.

## Prior Art
This is an attempt at a revival of 7771 and other discussions around storing and utilizing Metadata throughout Prometheus. 

Unlike the reference work, this only deals with appending metadata to the WAL, and aims to start a bigger discussion around how we can make Metadata more useful throughout Prometheus.

As such, reading from the WAL and sending over to remote_write is not in the scope of this PR and will be handled with a future pull request..

Older Mailing list discussion: https://groups.google.com/g/prometheus-developers/c/w-eotGenBGg/m/UZWdxzcBBwAJ

Older Design doc: https://docs.google.com/document/d/1LY8Im8UyIBn8e3LJ2jB-MoajXkfAqW2eKzY735aYxqo/edit#heading=h.bik9uwphqy3g

## Notes
After I get some feedback on the overall idea, I will extend the test suite to cover more cases and increase overall confidence on how the feature works and interacts with different pieces of Prometheus.

### CODE REVIEW

File: tsdb/head_append.go [hunk 1]
```diff
@@ -289,6 +309,15 @@ func (a *headAppender) Append(ref storage.SeriesRef, lset labels.Labels, t int64
 		V:   v,
 	})
 	a.sampleSeries = append(a.sampleSeries, s)
+	if hasNewMetadata {
+		a.metadata = append(a.metadata, record.RefMetadata{
```
Comment (rfratto) [inline tsdb/head_append.go:313]: a.metadata doesn't look like it ever gets cleared out after logging

File: tsdb/head_append.go [hunk 2]
```diff
@@ -265,7 +267,25 @@ func (a *headAppender) Append(ref storage.SeriesRef, lset labels.Labels, t int64
...
+		s.typ = meta.Type
+	}
+	if s.unit != meta.Unit {
+		hasNewMetadata = true
+		s.unit = meta.Unit
+	}
+	if s.help != meta.Help {
+		hasNewMetadata = true
+		s.help = meta.Help
+	}
```
Comment (rfratto) [inline tsdb/head_append.go:287]: Maybe the in-memory series should store the model/metadata type instead of keeping each field separately?

File: scrape/scrape.go [hunk 3]
```diff
@@ -1521,9 +1536,25 @@ loop:
...
+				// specified, they were treated as zeroed. We have to understand whether we actually need to use this approach
+				// and also take the same approach in the `if ok{` block above.
+				// if metaEntry.lastIterObservedType == currIter {
+				// if metaEntry.lastIterObservedUnit == currIter {
+				// if metaEntry.lastIterObservedHelp == currIter {
+				meta.Type = metaEntry.typ
+				meta.Unit = metaEntry.unit
+				meta.Help = metaEntry.help
+			}
+			sl.cache.metaMtx.Unlock()
```
Comment (rfratto) [inline scrape/scrape.go:1554]: Checking for changed metadata should be in the branch where the series was cached (i.e, the `if ok` branch, not in the else (i.e., it's a new series)

File: scrape/scrape.go [hunk 4]
```diff
@@ -1521,9 +1536,25 @@ loop:
...
 		}
 
-		ref, err = app.Append(ref, lset, t, v)
+		ref, err = app.Append(ref, lset, meta, t, v)
```
Comment (rfratto) [inline scrape/scrape.go:1557]: Hypothetically we only need to append the metadata if the metadata for a series has changed or it's a new series

File: scrape/scrape.go [hunk 5]
```diff
@@ -1663,6 +1695,43 @@ const (
 	scrapeBodySizeBytesMetricName = "scrape_body_size_bytes" + "\xff"
 )
 
+var (
```
Comment (rfratto) [inline scrape/scrape.go:1698]: Whether we want metadata for report samples should be a separate PR imo, and start with no metadata for now

**Review by rfratto — COMMENTED** [review]
Great work :)

File: scrape/scrape.go [hunk 6]
```diff
@@ -1451,6 +1461,7 @@ loop:
 		var (
 			et          textparse.Entry
 			sampleAdded bool
+			hasMetadata bool
```
Comment (rfratto) [inline scrape/scrape.go:1464]: nit: hasMetadata seems not totally accurate, since metadata can exist but still not be written. `shouldWriteMetadata`?

File: scrape/scrape.go [hunk 7]
```diff
@@ -1560,6 +1608,13 @@ loop:
 			e = exemplar.Exemplar{} // reset for next time round loop
 		}
 
+		if hasMetadata {
+			_, metadataErr := app.AppendMetadata(ref, meta)
+			if metadataErr != nil {
+				// No need to fail the scrape on errors appending metadata.
+				level.Debug(sl.l).Log("msg", "Error when appending metadata in scrape loop", "metadata", fmt.Sprintf("%+v", meta), "err", metadataErr)
```
Comment (rfratto) [inline scrape/scrape.go:1615]: nit: I'd add the ref as a field in the log too

File: tsdb/head_append.go [hunk 8]
```diff
@@ -65,6 +66,17 @@ func (a *initAppender) AppendExemplar(ref storage.SeriesRef, l labels.Labels, e
 	return a.app.AppendExemplar(ref, l, e)
 }
 
+func (a *initAppender) AppendMetadata(ref storage.SeriesRef, m metadata.Metadata) (storage.SeriesRef, error) {
+	if a.app != nil {
+		return a.app.AppendMetadata(ref, m)
+	}
+
+	// TODO check if we'll be keeping track of time here as well.
+	// a.head.initTime(t)
```
Comment (rfratto) [inline tsdb/head_append.go]: This is done for the other Appends, so I'd say it's probably expected to init the time here as well.

File: tsdb/head_append.go [hunk 9]
```diff
@@ -350,6 +379,46 @@ func (a *headAppender) AppendExemplar(ref storage.SeriesRef, lset labels.Labels,
...
+	// Get Series
+	s := a.head.series.getByID(chunks.HeadSeriesRef(ref))
+
+	// TODO: Do we need to add label set so we can get series by their hash?
+	// if s == nil {
+	// 	s = a.head.series.getByHash(lset.Hash(), lset)
+	// 	if s != nil {
+	// 		ref = storage.SeriesRef(s.ref)
+	// 	}
+	// }
```
Comment (rfratto) [inline tsdb/head_append.go:390]: I think you have to add the label set, yeah. ref, from what I understand, is an optional argument, and the caller might not always have one or save one. lset should be used as a fallback to ensure that the function can't fail if the caller isn't caching ref IDs.

File: tsdb/record/record.go [hunk 10]
```diff
@@ -60,6 +63,14 @@ type RefSample struct {
 	V   float64
 }
 
+// RefMetadata is the series metadata.
+type RefMetadata struct {
+	Ref  chunks.HeadSeriesRef
+	Type textparse.MetricType
```
[the remark this answers is not in the document]
Comment (rfratto) [reply tsdb/record/record.go:69]: (I'm fine with this being a string now honestly)

File: model/metadata/metadata.go [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+	Type textparse.MetricType
+	Unit string
+	Help string
+}
+
+var emptyMetadata = Metadata{}
+
+func EmptyMetadata() Metadata {
+	return emptyMetadata
+}
```
Comment (rfratto) [inline model/metadata/metadata.go:29]: tiny nit: I don't know if functions to return a zero value are really common, I'd just have the caller use `metadata.Metadata{}` if they needed it (which is also shorter than calling `metadata.EmptyMetadata()`.

Don't have a function for this, have the caller create the zero struct manually if they need it.

**Review by rfratto — APPROVED** [review]
I think this is ready for a PR upstream, honestly. nice work :)

File: model/metadata/metadata.go (hunk 11 above)
Comment (tpaschalis) [reply model/metadata/metadata.go:29]: I don't really like it too tbh, but I saw it being used in `EmptySeriesSet`.  Fortunately with this refactor it's only used in one place, so I'll get rid of it.
